tests/upgrade-tests/test_upgrade_charms.py

```diff
@@ -14,9 +14,10 @@
 @pytest.mark.parametrize('bundle,from_channel,to_channel', test_cases)
 async def test_upgrade_charms(bundle, from_channel, to_channel):
     async with temporary_model() as model:
-        await model.deploy(bundle, channel=from_channel)
+        await deploy_bundle(model, bundle, from_channel)
         await wait_for_ready(model)
         for app in model.applications.values():
             await app.upgrade_charm(channel=to_channel)
+        await add_new_e2e_relation(model)
```

Comment:
So this test will deploy the old e2e, upgrade it, then add the new relation.

Does it make sense to upgrade e2e at all? It seems like it'd be cleaner to just deploy the newer e2e after the upgrades have taken place -- that way, all of its relations are handled in one place. On the other hand, being able to deploy e2e in advance might time in the test run.

I'm not sure where I stand on that, so we'll call it good and I'll revisit later :)